_posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md

```diff
@@ -0,0 +1,105 @@
+- Getting a list of shredded types required having to use a separate function, `jsonifyWithInventory`.
+
+In 0.4.0, the `EventTransformer` API has been fully deprecated in favor of `Event` - a single typesafe container that contains all 132 members of a [canonical Snowplow event][canonical-event-model]. All fields are automatically converted to appropriate non-String types where possible - for instance, the `event_id` column is represented as a [UUID instance][java-uuid], while timestamps are converted into optional [Instant][java-instant] values, eliminating the need for common string conversions. Contexts and unstructured events are also wrapped in [self-describing data container types][self-describing-type], allowing for advanced operations such as Iglu URI lookups.
+
+The case class has the following primary functions:
+
+- `Event.parse(line)` - similar to the old `transform` function, this method accepts an enriched Snowplow event as a string and returns an `Event` instance as a result.
+- `event.toJson(lossy)` - similar to the old `getValidatedJsonEvent` function, it transforms an `Event` into a validated JSON whose keys are the field names corresponding to the EnrichedEvent POJO of the Scala Common Enrich project. If the lossy argument is true, any self-describing events in the fields (unstruct_event, contexts and derived_contexts) are returned in a "shredded" format, e.g. `"unstruct_event_com_acme_1_myField": "value"`. If it is set to false, they are not flattened into underscore-separated top-level fields, using a standard self-describing format instead.
+- `event.inventory` - extracts metadata from the event containing information about the types and Iglu URIs of its shred properties (unstruct_event, contexts and derived_contexts). Unlike version 0.3.0, it no longer requires a `transformWithInventory` call and can be obtained from any `Event` instance.
+- `atomic` - returns the event as a map of keys to Circe JSON values, while dropping inventory fields. This method can be used to modify an event's JSON AST before converting it into a final result.
+- `ordered` - returns the event as a list of key/Circe JSON value pairs. Unlike `atomic`, which has randomized key ordering, this method returns the keys in the order of the canonical event model, and is particularly useful for working with relational databases.
+
+<h2 id="typesafe-api">2. Using the typesafe API</h2>
+
+Since base results of the Scala Analytics SDK are now members of the `Event` case class, their output needs to be converted to JSON strings. For instance, the following code can be used in a data modelling job to load a series of events into a Spark dataframe:
+
+{% highlight scala %}
+import com.snowplowanalytics.snowplow.analytics.scalasdk.Event
+
+val events = input
+  .map(line => Event.parse(line))
+  .filter(_.isValid)
```

Comment:
Can be simplified to:

```
val events = input
  .map(line => Event.parse(line))
  .flatMap(_.toOption)
  .map(event => event.toJson(true).noSpaces)
```

We don't need filter first - as we're throwing away this information anyway.